### ISSUE TITLE
Why does node UMLS:C0032961 have biolink category label "named thing"?

### ISSUE THREAD

**Issue by saramsey** [RTXteam/RTX#1219]
Try this Cypher on kg2endpoint-kg2-5-1.rtx.ai:7474:
```
match (n {id: 'UMLS:C0032961'}) return n.id, n.name, n.category, n.description;
```
<img width="926" alt="Screen Shot 2021-02-02 at 3 35 17 PM" src="https://user-images.githubusercontent.com/5562107/106676731-4ac17d80-656c-11eb-8a43-28e5424183a6.png">

This node has UMLS semantic type T040:
```
UMLS Semantic Type: UMLS_STY:T040; UMLS Semantic Type: UMLS_STY:T040"
```
so one would naively imagine that this node would be given the category `physiological process`, per this line in `curies-to-categories.yaml`:
```
    UMLS_STY:T040: physiological process
```

**Comment by ericawood** [comment]
This is certainly a problem! However, I think that line in `curies-to-categories.yaml` is also incorrect:
From curies-to-categories.yaml:
`UMLS_STY:T040: physiological process`

However, from UMLS's website:
![image](https://user-images.githubusercontent.com/36611732/106924764-44deb000-66c4-11eb-9cf7-15ffe449ac9f.png)

And from the Biolink YAML file:
![image](https://user-images.githubusercontent.com/36611732/106925301-d0f0d780-66c4-11eb-96cb-37f50b54ac92.png)

@saramsey Is this mismatch an issue?

**Comment by ericawood** [comment]
This is a way more problematic issue than we originally thought (it is plaguing far more nodes than originally thought and likely the reason that #762 appears to be an issue).

For reference, `UMLS:C0032961` is in the following TTL files:
- `umls-go.ttl`
- `umls-lnc.ttl`
- `umls-mdr.ttl`
- `umls-medlineplus.ttl`
- `umls-msh.ttl`
- `umls-mth.ttl`
- `umls-nci.ttl`
- `umls-psy.ttl`
- `umls-snomedct_us.ttl`

This means that each of the `UMLS:C0032961` nodes created from those files must be merged.

In `multi_ont_to_json_kg.py`, `merge_two_dicts` from `kg2_util.py` is called:
```
                        if cui_node_dict_existing is not None:
                            cui_node_dict = kg2_util.merge_two_dicts(cui_node_dict,
                                                                     cui_node_dict_existing,
                                                                     biolink_depth_getter)
```

I printed out the input `cui_node_dict` as "Other Node Dictionary", `cui_node_dict_existing` as "Existing Node Dictionary", and the output `cui_node_dict` as "CUI Node Dict". The way that `merge_two_dicts` is supposed to work is to take whichever category is most specific. This is not what happens.


The problem lies in line 471 of `kg2_util.py`, within the `merge_two_dicts` function.
```
                            if biolink_depth_getter is not None:
                                depth_x = biolink_depth_getter(convert_snake_case_to_camel_case(stored_value, uppercase_first_letter=True))
                                depth_y = biolink_depth_getter(convert_snake_case_to_camel_case(value, uppercase_first_letter=True))
                                if depth_y is not None:
                                    if depth_x is not None:
                                        if depth_y > depth_x:
                                            ret_dict[key] = value
                                    else:
                                        ret_dict[key] = value
```
If depth_x is larger, than ret_dict[key] should store stored_value NOT value. Essentially, if they are in the wrong order, the node gets the simpler category. This is problematic and causes issues in more nodes than just `UMLS:C0032961`. Here's a non `UMLS:C0032961` example:
```
Existing Node Dictionary: {
        "category": "biolink:MolecularActivity",
        "category_label": "molecular_activity",
        "creation_date": null,
        "deprecated": false,
        "description": "Any process that stops, prevents or reduces the frequency, rate or extent of L-glutamate import into a cell. [GOC:TermGenie]",
        "full_name": null,
        "id": "UMLS:C4689689",
        "iri": "https://identifiers.org/umls:C4689689",
        "name": "negative regulation of L-glutamate import across plasma membrane",
        "ontology node ids": [],
        "provided_by": "identifiers_org_registry:umls",
        "publications": [],
        "replaced_by": null,
        "synonym": [],
        "update_date": "2019",
        "xrefs": []
    }
    Other Node Dictionary: {
        "category": "biolink:NamedThing",
        "category_label": "named_thing",
        "creation_date": null,
        "deprecated": false,
        "description": "Any process that modulates the frequency, rate or extent of L-glutamate import into a cell. [GOC:TermGenie]",
        "full_name": null,
        "id": "UMLS:C4689689",
        "iri": "https://identifiers.org/umls:C4689689",
        "name": "regulation of L-glutamate import across plasma membrane",
        "ontology node ids": [],
        "provided_by": "identifiers_org_registry:umls",
        "publications": [],
        "replaced_by": null,
        "synonym": [],
        "update_date": "2019",
        "xrefs": []
    }
    CUI Node Dict: {
        "category": "biolink:NamedThing",
        "category_label": "named_thing",
        "creation_date": null,
        "deprecated": false,
        "description": "Any process that modulates the frequency, rate or extent of L-glutamate import into a cell. [GOC:TermGenie]; Any process that stops, prevents or reduces the frequency, rate or extent of L-glutamate imp        ort into a cell. [GOC:TermGenie]",
        "full_name": null,
        "id": "UMLS:C4689689",
        "iri": "https://identifiers.org/umls:C4689689",
        "name": "negative regulation of L-glutamate import across plasma membrane",
        "ontology node ids": [],
        "provided_by": "identifiers_org_registry:umls",
        "publications": [],
        "replaced_by": null,
        "synonym": [],
        "update_date": "2019",
        "xrefs": []
    }
```

I'm currently running some tests related to this on `kg2lindsey` and will report back based on those.

**Comment by ericawood** [comment]
Also, I know that the TUIs are being read properly because I ran `multi_ont_to_json_kg.py` on just `umls-msh.ttl` and node `UMLS:C0032961` had the correct category label.

**Comment by ericawood** [comment]
Here's the `UMLS:C0032961` example:
```
 Existing Node Dictionary: {
     "category": "biolink:PhysiologicalProcess",
     "category_label": "physiological_process",
     "creation_date": null,
     "deprecated": false,
     "description": "The state or condition of having a developing embryo or fetus in the body (uterus), after union of an ovum and spermatozoon, during the period from conception to birth.; The status during which female          mammals carry their developing young (EMBRYOS or FETUSES) in utero before birth, beginning from FERTILIZATION to BIRTH.; The set of physiological processes that allow an embryo or foetus to develop within the body of a          female animal. It covers the time from fertilization of a female ovum by a male spermatozoon until birth. [ISBN:0192800825]; You're going to have a baby! It's an exciting time, but it can also feel a bit overwhelming. Yo         u may have a lot of questions, including what you can do to give your baby a healthy start. To keep you and your baby healthy during pregnancy, it is important to  Have regular visits with your health care provider. Thes         e prenatal care visits help make sure that you and your baby are healthy. And if there are any health problems, your provider can find them early. Getting treatment right away can cure many problems and prevent others. E         at healthy and drink plenty of water. Good nutrition during pregnancy includes eating a variety of  Fruits Vegetables Whole grains Lean meats or other protein sources Low-fat dairy products  Take prenatal vitamins. Pregn         ant women need higher amounts of certain vitamins and minerals, such as folic acid and iron. Be careful with medicines. Always check with your health care provider before you start or stop any medicine. This includes ove         r-the-counter medicines and dietary or herbal supplements. Stay active. Physical activity can help you stay strong, feel and sleep better, and prepare your body for birth. Check with your provider about which types of ac         tivities are right for you. Avoid substances that could hurt your baby, such as alcohol, drugs, and tobacco.  Your body will keep changing as your baby grows. It can be hard to know whether a new symptom is normal or cou         ld be a sign of a problem. Check with your health care provider if something is bothering or worrying you.",
     "full_name": null,
     "id": "UMLS:C0032961",
     "iri": "https://identifiers.org/umls:C0032961",
     "name": "Pregnancy", 
     "ontology node ids": [
         "http://purl.bioontology.org/ontology/MEDLINEPLUS/C0032961",
         "http://purl.bioontology.org/ontology/MTH/C0032961"
     ],
     "provided_by": "identifiers_org_registry:umls",
     "publications": [
         "ISBN:0192800825"
     ],
     "replaced_by": null,
     "synonym": [], 
     "update_date": "2017",
     "xrefs": []
 }
 Other Node Dictionary: {
     "category": "biolink:NamedThing",
     "category_label": "named_thing",
     "creation_date": null,
     "deprecated": false,
     "description": "Patient currently pregnant",
     "full_name": null,
     "id": "UMLS:C0032961",
     "iri": "https://identifiers.org/umls:C0032961",
     "name": "Pregnant",
     "ontology node ids": [],
     "provided_by": "identifiers_org_registry:umls",
     "publications": [],
     "replaced_by": null,
     "synonym": [],
     "update_date": "2017",
     "xrefs": []
 }
 CUI Node Dict: {
     "category": "biolink:NamedThing",
     "category_label": "named_thing",
     "creation_date": null,
     "deprecated": false,
     "description": "Patient currently pregnant; The state or condition of having a developing embryo or fetus in the body (uterus), after union of an ovum and spermatozoon, during the period from conception to birth.; Th         e status during which female mammals carry their developing young (EMBRYOS or FETUSES) in utero before birth, beginning from FERTILIZATION to BIRTH.; The set of physiological processes that allow an embryo or foetus to d         evelop within the body of a female animal. It covers the time from fertilization of a female ovum by a male spermatozoon until birth. [ISBN:0192800825]; You're going to have a baby! It's an exciting time, but it can also          feel a bit overwhelming. You may have a lot of questions, including what you can do to give your baby a healthy start. To keep you and your baby healthy during pregnancy, it is important to  Have regular visits with you         r health care provider. These prenatal care visits help make sure that you and your baby are healthy. And if there are any health problems, your provider can find them early. Getting treatment right away can cure many pr         oblems and prevent others. Eat healthy and drink plenty of water. Good nutrition during pregnancy includes eating a variety of  Fruits Vegetables Whole grains Lean meats or other protein sources Low-fat dairy products  T         ake prenatal vitamins. Pregnant women need higher amounts of certain vitamins and minerals, such as folic acid and iron. Be careful with medicines. Always check with your health care provider before you start or stop any          medicine. This includes over-the-counter medicines and dietary or herbal supplements. Stay active. Physical activity can help you stay strong, feel and sleep better, and prepare your body for birth. Check with your prov         ider about which types of activities are right for you. Avoid substances that could hurt your baby, such as alcohol, drugs, and tobacco.  Your body will keep changing as your baby grows. It can be hard to know whether a          new symptom is normal or could be a sign of a problem. Check with your health care provider if something is bothering or worrying you.",
     "full_name": null,
     "id": "UMLS:C0032961",
     "iri": "https://identifiers.org/umls:C0032961",
     "name": "Pregnancy",
     "ontology node ids": [
         "http://purl.bioontology.org/ontology/MEDLINEPLUS/C0032961",
         "http://purl.bioontology.org/ontology/MTH/C0032961"
     ],
     "provided_by": "identifiers_org_registry:umls",
     "publications": [
         "ISBN:0192800825"
     ],
     "replaced_by": null,
     "synonym": [],
     "update_date": "2017",
     "xrefs": []
 }
```

**Comment by ericawood** [comment]
My above comments are not the correct diagnosis. The actual problem is that `depth_x` and `depth_y` are both None because `biolink_depth_getter` calls `get_biolink_categories_ontology_depths`, which has a return dictionary where all of the CamelCased categories are prefixed with `biolink:`. Hence, the issue is that `stored_value` and `value` (for `category_label`) do not have the `biolink:` prefix. Here is the problem code:

```
                        elif key == 'category_label':
                            if biolink_depth_getter is not None:
                                depth_x = biolink_depth_getter(convert_snake_case_to_camel_case(stored_value, uppercase_first_letter=True))
                                depth_y = biolink_depth_getter(convert_snake_case_to_camel_case(value, uppercase_first_letter=True))
                                if depth_y is not None:
                                    if depth_x is not None:
                                        if depth_y > depth_x:
                                            ret_dict[key] = value
```

Here is rough draft working code (I will formalize this later with `CURIE_PREFIX_BIOLINK`):
```
                         elif key == 'category_label':
                             if biolink_depth_getter is not None:
                                 depth_x = biolink_depth_getter('biolink:' + convert_snake_case_to_camel_case(stored_value, uppercase_first_letter=True))
                                 depth_y = biolink_depth_getter('biolink:' + convert_snake_case_to_camel_case(value, uppercase_first_letter=True))
                                 if depth_y is not None:
                                     if depth_x is not None:
                                         if depth_y > depth_x:
                                             ret_dict[key] = value

```

Regarding `category` (not `category_label`), the problem code is:
```
                        elif key == 'category':
                            if biolink_depth_getter is not None:
                                value_category = value.replace(CURIE_PREFIX_BIOLINK + ':', '')
                                stored_value_category = stored_value.replace(CURIE_PREFIX_BIOLINK + ':', '')
                                depth_x = biolink_depth_getter(stored_value_category)
                                depth_y = biolink_depth_getter(value_category)
                                if depth_y is not None:
                                    if depth_x is not None:
                                        if depth_y > depth_x:
                                            ret_dict[key] = value
```

Here is working code for that section:
```
                         elif key == 'category':
                             if biolink_depth_getter is not None:
                                 depth_x = biolink_depth_getter(stored_value)
                                 depth_y = biolink_depth_getter(value)
                                 if depth_y is not None:
                                     if depth_x is not None:
                                         if depth_y > depth_x:
                                             ret_dict[key] = value

```

@saramsey Do you have any thoughts about this?

**Comment by saramsey** [comment]
Thank you @ericawood. Very nice detective work!  This fix looks sensible. Let's try a test build with it.

**Comment by ericawood** [comment]
Here's the results from the test build:
![image](https://user-images.githubusercontent.com/36611732/107690836-7ffe5780-6c5f-11eb-8451-4bacb4409751.png)

![image](https://user-images.githubusercontent.com/36611732/107690901-986e7200-6c5f-11eb-9bb8-5a94c098f42c.png)

Here's the results from the PREVIOUS test build (pre-fix):
![image](https://user-images.githubusercontent.com/36611732/107691088-de2b3a80-6c5f-11eb-86e8-14845145781e.png)

![image](https://user-images.githubusercontent.com/36611732/107691006-c2279900-6c5f-11eb-87a4-ca6a0b98d2a5.png)

In summary, the number of UMLS `named_thing` nodes dropped by 8101 (a 65% drop). I am unsure why the number of `named_thing` nodes only dropped by 5602 overall (a 37% drop) and will investigate that further.

**Comment by ericawood** [comment]
> In summary, the number of UMLS `named_thing` nodes dropped by 8101 (a 65% drop). I am unsure why the number of `named_thing` nodes only dropped by 5602 overall (a 37% drop) and will investigate that further.

The lack of a larger overall drop is largely due to `go-plus.owl` (an increase from 2627 to 5023) and `MEDDRA` (an increase from 2 to 102.

**Comment by ericawood** [comment]
Based on two full builds of `multi_ont_to_json_kg.py` using `ont-load-inventory.yaml`, one with the old code and one with the fix, I discovered. Here's the data from it:
Source | Category Label | Old Code Count | New Code Count | Percent Decrease
-- | -- | -- | -- | --
Total | named_thing | 442499 | 403917 | 8.719115749
EFO:efo.owl | named_thing | 2643 | 2615 | 1.059402194
OBO:bfo.owl | named_thing | 31 | 31 | 0
OBO:chebi.owl | named_thing | 18498 | 18484 | 0.07568385771
OBO:cl.owl | named_thing | 1 | 1 | 0
OBO:doid.owl | named_thing | 44 | 44 | 0
OBO:ehdaa2.owl | named_thing | 21 | 21 | 0
OBO:foodon.owl | named_thing | 2113 | 2079 | 1.609086607
OBO:genepio.owl | named_thing | 469 | 469 | 0
go-plus.owl | named_thing | 4998 | 4963 | 0.700280112
hp.owl | named_thing | 245 | 245 | 0
ino.owl | named_thing | 2 | 2 | 0
mondo.owl | named_thing | 2146 | 2112 | 1.584342964
ro.owl | named_thing | 1 | 1 | 0
uberon/ext.owl | named_thing | 10 | 5 | 50
ORPHANET | named_thing | 749 | 749 | 0
UMLS_STY | named_thing | 1 | 11 | -1000
identifiers_org_registry:umls | named_thing | 105013 | 68628 | 34.64809119
umls_source:ATC | named_thing | 1427 | 1427 | 0
umls_source:DRUGBANK | named_thing | 22 | 22 | 0
umls_source:FMA | named_thing | 170 | 170 | 0
umls_source:HL7 | named_thing | 5251 | 3705 | 29.44201105
umls_source:HPO | named_thing | 1 | 1 | 0
umls_source:ICD10PCS | named_thing | 189499 | 189499 | 0
umls_source:ICD9CM | named_thing | 22386 | 22386 | 0
umls_source:MED-RT | named_thing | 672 | 291 | 56.69642857
umls_source:MEDDRA | named_thing | 102 | 102 | 0
umls_source:MEDLINEPLUS | named_thing | 201 | 174 | 13.43283582
umls_source:MSH | named_thing | 15243 | 15238 | 0.03280194187
umls_source:MTH | named_thing | 38967 | 38967 | 0
umls_source:NCI | named_thing | 19755 | 19667 | 0.4454568464
umls_source:NDDF | named_thing | 1455 | 1455 | 0
umls_source:PDQ | named_thing | 3717 | 3717 | 0
umls_source:PSY | named_thing | 5234 | 5234 | 0
umls_source:SNOMEDCT | named_thing | 640 | 640 | 0
umls_source:VANDF | named_thing | 772 | 772 | 0


Since it was successful, I am committing my changes. The increase in `UMLS_STY` `named_thing` nodes will be addressed in #1263.

**Comment by ericawood** [comment]
This is fixed in KG2.5.2:

```
{
  "iri": "https://identifiers.org/umls:C0032961",
  "category_label": "physiological_process",
  "deprecated": "False",
  "name": "Pregnancy",
  "description": "Patient currently pregnant; The state or condition of having a developing embryo or fetus in the body (uterus), after union of an ovum and spermatozoon, during the period from conception to birth.; The status during which female mammals carry their developing young (EMBRYOS or FETUSES) in utero before birth, beginning from FERTILIZATION to BIRTH.; The set of physiological processes that allow an embryo or foetus to develop within the body of a female animal. It covers the time from fertilization of a female ovum by a male spermatozoon until birth. [ISBN:0192800825]; You're going to have a baby! It's an exciting time, but it can also feel a bit overwhelming. You may have a lot of questions, including what you can do to give your baby a healthy start. To keep you and your baby healthy during pregnancy, it is important to  Have regular visits with your health care provider. These prenatal care visits help make sure that you and your baby are healthy. And if there are any health problems, your provider can find them early. Getting treatment right away can cure many problems and prevent others. Eat healthy and drink plenty of water. Good nutrition during pregnancy includes eating a variety of  Fruits Vegetables Whole grains Lean meats or other protein sources Low-fat dairy products  Take prenatal vitamins. Pregnant women need higher amounts of certain vitamins and minerals, such as folic acid and iron. Be careful with medicines. Always check with your health care provider before you start or stop any medicine. This includes over-the-counter medicines and dietary or herbal supplements. Stay active. Physical activity can help you stay strong, feel and sleep better, and prepare your body for birth. Check with your provider about which types of activities are right for you. Avoid substances that could hurt your baby, such as alcohol, drugs, and tobacco.  Your body will keep changing as your baby grows. It can be hard to know whether a new symptom is normal or could be a sign of a problem. Check with your health care provider if something is bothering or worrying you.; UMLS Semantic Type: UMLS_STY:T040; UMLS Semantic Type: UMLS_STY:T040",
  "provided_by": "identifiers_org_registry:umls",
  "id": "UMLS:C0032961",
  "category": "biolink:PhysiologicalProcess",
  "update_date": "2017",
  "publications": [
    "ISBN:0192800825"
  ]
}
```
